sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala

```diff
@@ -33,11 +33,18 @@ object TypeUtils {
   }
 
   def checkForOrderingExpr(t: DataType, caller: String): TypeCheckResult = {
-    if (t.isInstanceOf[AtomicType] || t == NullType) {
-      TypeCheckResult.TypeCheckSuccess
-    } else {
-      TypeCheckResult.TypeCheckFailure(s"$caller accepts non-complex types, not $t")
+    t match {
+      case i: AtomicType => TypeCheckResult.TypeCheckSuccess
+      case n: NullType => TypeCheckResult.TypeCheckSuccess
+      case s: StructType =>
+        if (s.fields.exists(_.dataType.isInstanceOf[AtomicType])) {
```

Comment:
this should do a recursive check.

I think it might be easier if you create a supportOrdering function for StructType, and then uses that to determine if we support ordering here.